src/base58.rs

```diff
@@ -0,0 +1,182 @@
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt::{Display, Formatter},
+    ops::{Deref, DerefMut},
+    str::FromStr,
+};
+
+#[serde_with::serde_as]
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize, Hash, PartialOrd, Ord)]
+#[serde(transparent)]
+pub struct AsString<T>(#[serde_as(as = "serde_with::DisplayFromStr")] pub T)
+where
+    T: Display + FromStr,
+    <T as FromStr>::Err: Display;
+
+impl<T> Deref for AsString<T>
+where
+    T: Display + FromStr,
+    <T as FromStr>::Err: Display,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for AsString<T>
+where
+    T: Display + FromStr,
+    <T as FromStr>::Err: Display,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, PartialOrd, Ord)]
+pub struct Base58<T>(pub T);
+
+impl<T> Deref for Base58<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for Base58<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T: AsRef<[u8]>> Display for Base58<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        bs58::encode(self.0.as_ref()).into_string().fmt(f)
+    }
+}
+
+impl<T, E> FromStr for Base58<T>
+where
+    Base58<T>: for<'a> TryFrom<&'a [u8], Error = E>,
+    E: From<bs58::decode::Error>,
```

Comment:
I can't implement `TryFrom<&[u8]>` for foreign types like `PublicKey` from ed25519-dalek, hence I implement it on `Base58<PublicKey>`. Here I decode base58 string and depend on `TryFrom<&[u8]>` impl to further decode data to `Base58<T>`.